addons/crm/static/src/js/forecast/forecast_models.js

```diff
@@ -0,0 +1,163 @@
+        if (lastGroup) {
+            lastGroup = this.localData[lastGroup.id];
+            forecast.end = moment(lastGroup.range[this.forecast_field].to);
+        }
+        return this.handle;
+    },
+    /**
+     * In order to display the sample data, the context and domain need to be the same as before a reload.
+     * If the user switches to another view (list, pivot, ...) and comes back without adding any record or modifying the
+     * search bar, we still want to display the sample data, but the actual domain and context won't be the same because
+     * load and reload are modifying them.
+     * 
+     * As such, we will only compare the original domain and context (without the use of the ForecastService) to determine
+     * whether the sample data should be enabled or not. We also need to check if the groupby changed as in the original function
+     * @see AbtractModel._haveParamsChanged
+     */
+    __haveParamsChanged(params) {
+        var originalParams = [this.baseContext, this.baseDomain];
+        var currentParams = [params.context, params.domain];
+        if (JSON.stringify(originalParams) !== JSON.stringify(currentParams)) {
+            return true;
+        }
+        if (this.useSampleModel && 'groupBy' in params) {
+            return JSON.stringify(params.groupBy) !== JSON.stringify(this.loadParams.groupedBy);
+        }
+    },
+    /**
+     * Whether the sample data should be displayed (false) or not (true). This had to be computed beforehand,
+     * because `this.__reload` is actually changing the context and the domain, but this function is evaluated
+     * way later @see AbstractModel.reload
+     * @see __haveParamsChanged
+     * @override
+     */
+    _haveParamsChanged() {
+        return this.haveParamsChanged;
+    },
+    /**
+     * Update the domain, context and forecast_field if needed.
+     * 
+     * haveParamsChanged will be used to determine if the sample data should be
+     * displayed or not
+     * 
+     * @param {Object} params
+     * @param {string[]} [params.domain] optional
+     * @param {Object} [params.context] optional
+     */
+    _updateParams(params) {
+        if ("domain" in params && "context" in params) {
+            this.haveParamsChanged = this.__haveParamsChanged(params);
+        }
+        if ("domain" in params) {
+            this.baseDomain = params.domain;
+        }
+        if ("context" in params) {
+            this.baseContext = params.context;
+            this.forecast_field = params.context.forecast_field;
+        }
+    },
+    /**
+     * Applies the forecast logic to the domain and context if needed
+     * before the read_group.
+     * 
+     * After the read_group, checks the end date of the last displayed group
+     * in order to be able to add the next group with the ForecastColumnQuickCreate widget
+     * @override
+     */
+    async __reload(id, params) {
```

Comment:
Couldn't some computation be extracted from __load ? Seems you have duplicated code, notably from using TimingForecastFillTemporalGroupingAsynchronousMasterService (naming may update).